### PR TITLE
CT-656 - Make explainTransaction methods return promise

### PR DESCRIPTION
**Pull request by CapnMigraine**
The explainTransaction method for EOS makes an asynchronous call and returns a promise. In order to be consistent, other coins should return a promise as well.

References:
https://bitgoinc.atlassian.net/browse/CT-656
https://bitgoinc.atlassian.net/browse/CT-637

### CODE REVIEW

File: modules/core/src/v2/coins/abstractUtxoCoin.ts [hunk 1]
```diff
@@ -13,6 +13,38 @@ const errors = require('../../errors');
...
+  amount: string;
+}
+
+export interface TransactionFee {
+  fee: number;
+  feeRate?: number;
+  size: number
+}
+
+export interface TransactionExplanation {
```
Comment (marktoda) [inline modules/core/src/v2/coins/abstractUtxoCoin.ts]: Is this the same interface across coins? Maybe it could be moved to baseCoin

**Comment by dannybitgo** [comment]
@CapnMigraine This is a breaking change, and would break the OVC: https://github.com/BitGo/offline-signing-tool/blob/f8341f88fba9b2ecb10125909a2520aa563d5e7a/frontend/scripts/signTransaction/signTransactionController.js#L150

**Comment by dannybitgo** [comment]
Does EOS need to return a promise for this? Can we find a way to wrap it so that doesn't happen instead?

**Comment by CapnMigraine** [comment]
> @CapnMigraine This is a breaking change, and would break the OVC: https://github.com/BitGo/offline-signing-tool/blob/f8341f88fba9b2ecb10125909a2520aa563d5e7a/frontend/scripts/signTransaction/signTransactionController.js#L150

Yeah, I'm working on changes to OVC to support this change

File: modules/core/src/v2/coins/abstractUtxoCoin.ts [hunk 2]
```diff
@@ -1006,126 +1038,128 @@ export class AbstractUtxoCoin extends BaseCoin {
     return areAllSignaturesValid;
   }
 
-  explainTransaction(params = {} as any) {
-    const { txHex, txInfo } = params;
+  explainTransaction(params: ExplainTransactionOptions): Bluebird<TransactionExplanation> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/abstractUtxoCoin.ts]: This should also take an optional `NodeCallback<TransactionExplanation>`, then pass to `asCallback` on the coroutine. Same for all the other implementations.

File: modules/core/src/v2/coins/xrp.ts [hunk 3]
```diff
@@ -376,40 +376,42 @@ export class Xrp extends BaseCoin {
...
-      txHex = params.txHex;
-    } catch (e) {
+  explainTransaction(params: ExplainTransactionOptions): Bluebird<TransactionExplanation> {
+    return co(function *() {
+      let transaction;
+      let txHex;
       try {
-        transaction = JSON.parse(params.txHex);
-        txHex = rippleBinaryCodec.encode(transaction);
+        transaction = rippleBinaryCodec.decode(params.txHex);
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/xrp.ts]: Why is this change here?

File: modules/core/src/v2/coins/xrp.ts [hunk 4]
```diff
@@ -485,32 +487,32 @@ export class Xrp extends BaseCoin {
    * @param callback
    */
   public recover(params: RecoveryOptions, callback): Bluebird<RecoveryInfo | string> {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/xrp.ts]: callback param should be `callback?: NodeCallback<RecoveryInfo | string>`

File: modules/core/test/v2/unit/coins/btc.ts [hunk 5]
```diff
@@ -413,36 +413,48 @@ describe('BTC:', function() {
...
-        it('should fail for invalid transaction hexes', () => {
-          (() => coin.explainTransaction())
-          .should.throw('invalid transaction hex, must be a valid hex string');
-
-          (() => coin.explainTransaction({
-            txHex: ''
-          })).should.throw('invalid transaction hex, must be a valid hex string');
+        it('should fail for invalid transaction hexes', co(function *() {
+          try {
+            yield coin.explainTransaction();
```
Comment (tylerlevine) [inline modules/core/test/v2/unit/coins/btc.ts]: `yield coin.explainTransaction().should.be.rejectedWith('invalid ...');` is a little cleaner

File: modules/core/src/v2/coins/abstractUtxoCoin.ts (hunk 1 above)
Comment (CapnMigraine) [reply modules/core/src/v2/coins/abstractUtxoCoin.ts]: It's not the same, but I created one in baseCoin that's extended by each coin that implements explainTransaction.

File: modules/core/src/v2/coins/xrp.ts (hunk 3 above)
Comment (CapnMigraine) [reply modules/core/src/v2/coins/xrp.ts]: git is showing it more complicated than it is... this try/catch matches the one in line 382 of the old version, the next one in line 387 is the same as the one in line 386 of the old version. Only the indentation changed.

File: modules/core/src/v2/coins/xrp.ts (hunk 4 above)
Comment (CapnMigraine) [reply modules/core/src/v2/coins/xrp.ts]: ok, fixed

File: modules/core/test/v2/unit/coins/btc.ts (hunk 5 above)
Comment (CapnMigraine) [reply modules/core/test/v2/unit/coins/btc.ts]: good one, changed.

File: modules/core/src/v2/coins/eos.ts [hunk 6]
```diff
@@ -82,16 +86,7 @@ interface DeserializedEosTransaction extends EosTransactionHeaders {
...
-
-interface TransactionExplanation {
-  displayOrder: string[];
-  id: string;
-  outputs: Recipient[];
-  changeOutputs: Recipient[];
-  outputAmount: string;
-  changeAmount: number;
-  fee: {};
+  memo?: string;
```
Comment (CapnMigraine) [inline modules/core/src/v2/coins/eos.ts]: small non-related fix, memo was not present in result

File: modules/core/src/v2/coins/eos.ts [hunk 7]
```diff
@@ -365,6 +360,7 @@ export class Eos extends BaseCoin {
       const transferActionData = EosJs.modules.Fcbuffer.fromBuffer(transferStruct, serializedTransferDataBuffer);
       transaction.address = transferActionData.to;
       transaction.amount = this.bigUnitsToBaseUnits(transferActionData.quantity.split(' ')[0]);
+      transaction.memo = transferActionData.memo;
```
Comment (CapnMigraine) [inline modules/core/src/v2/coins/eos.ts]: related to non-related fix up in line 89

File: modules/core/src/v2/coins/algo.ts [hunk 8]
```diff
@@ -49,34 +44,25 @@ export interface TransactionPrebuild {
...
+interface TransactionExplanation extends BaseCoinTransactionExplanation {
+  memo: string;
 }
 
 interface KeyPair {
   pub: string;
   prv: string;
 }
 
-const MAX_ALGORAND_NOTE_LENGTH = 1024;
```
Comment (CapnMigraine) [inline modules/core/src/v2/coins/algo.ts]: removed this because it's not used
